### DISCUSSION TITLE
Add signature to published images

### DISCUSSION THREAD

**Discussion by ivanvaccari** [nginx-proxy/nginx-proxy#2335] [Ideas]
I plan to use this this image to serve some services on a bare metal server.
The customer however requires signature verification for the installed softwares, and this image does not provide one.

By running
``` docker trust inspect nginxproxy/nginx-proxy```

the output is 

```no signatures or cannot access nginxproxy/nginx-proxy```

With this discussion i'm requesting to sign the images published on registries. Signing images should not (hopefully) introduce errors since is purely an extra security layer, and users pulling the image now for sure don't check signature, having it should not introduce problems.

**Comment by buchdag** [comment]
Hi @ivanvaccari 

I'm not familiar with this topic at all, it sound interesting but rather complex, do you have some recommended reads about it (other than the Docker documentation) ?

**Comment by SchoNie** [comment]
Somebody referenced this Github Action to me: https://github.com/sudo-bot/action-docker-sign
Maybe we can try this out as a separate gh action to an extra dedicated container image tag.

**Comment by ivanvaccari** [comment]
Hi, i'm not super familiar too since this is the first time also for me, bui i got it quite easily using the docs: https://docs.docker.com/engine/security/trust/

Not entering details (they're written off in docs), but a quick example for a demo image i've signed:
```sh
# one time, if you don't already have it. Asks for passkeys, create new ones and do not lose them
$> docker trust key generate ivanvaccarimitech

# Add a signer to notary server (docker hub in this case)
$> docker trust signer add --key .\ivanvaccarimitech.pub ivanvaccarimitech "ivanvaccarimitech/helloworld"

# Effectively sign data
$> docker trust sign ivanvaccarimitech/helloworld:1.0.0
```



After this, the inspect command returns some data:

```
docker trust inspect ivanvaccarimitech/helloworld
[
    {
        "Name": "ivanvaccarimitech/helloworld",
        "SignedTags": [
            {
                "SignedTag": "1.0.0",
                "Digest": "a3c9ad0af6880c32ddbaa03bdfe17bea478ec5e6892c1379d97550c715abad63",
                "Signers": [
                    "ivanvaccarimitech"
                ]
            }
        ],
        ........
]
```

**Comment by buchdag** [comment]
@ivanvaccari I've since played a bit with DCT. Signing single arch image is really easy but signing multi arch image is a nightmare. 

Unlike vanilla `docker build` command, there is no builtin DCT support inside buildx (https://github.com/docker/buildx/issues/313), so you have to manually sign each individual image then sign the manifest as a whole. This means using either the [notary CLI](https://github.com/notaryproject/notary/blob/master/docs/getting_started.md) (which is notoriously user unfriendly) or [`sudo-bot/action-docker-sign`](https://github.com/sudo-bot/action-docker-sign) (I'll let you see by yourself how convoluted [the manifest example](https://github.com/sudo-bot/action-docker-sign?tab=readme-ov-file#example-with-docker-buildx-using-manifests) is, and our own workflow would be way more complex than that because we build two images simultaneously, with multiple tags, and we push them on multiple registries 😞 ).

DockerHub library image are multiarch **and** signed but there is zero publicly available info on how they achieve this.

Unless there is a positive evolution to https://github.com/docker/buildx/issues/313, I don't see how we can realistically sign our multi arch images and regular tags. I don't think I'll give `sudo-bot/action-docker-sign` a try because they don't even bother to do tagged releases of their Github action and that's a no go for me.

It might however be possible, like @SchoNie suggested, to sign single arch images and make them available on separate tags, like `nginxproxy/nginx-proxy:1.5-signed` or `nginxproxy/nginx-proxy:1.5-amd64-signed` for instance.

**Comment by SchoNie** [comment]
I am testing with the [attest-build-provenance](https://github.com/actions/attest-build-provenance) Github action. It's different than DCT but the output is kinda the same creating a verifiable signature of artifacts created by actions.
Sample: [attested image created by workflow run](https://github.com/SchoNie/nginx-proxy/attestations/4370625) and signature in the public sigstore: [Sig 162019808](https://search.sigstore.dev/?logIndex=162019808)
And it's quite easy to implement.

workflow for the container image:
```
      - name: Attest
        uses: actions/attest-build-provenance@v2
        id: docker_attest
        with:
          subject-name: ${{ steps.docker_meta.outputs.tags }}
          subject-digest: ${{ steps.docker_build.outputs.digest }}
          push-to-registry: true
```

And for docker-gen attested files in the release assets as [multiple subjects (all release assets) in 1 attestation.](https://github.com/SchoNie/docker-gen/attestations/4371643)
If those are attested we could verify that in the nginx-proxy build workflow to improve security.
```
      - name: Attest release assets
        uses: actions/attest-build-provenance@v2
        with:
          subject-path: './docker-gen-*.tar.gz'
```

@buchdag Let me know if you are interested and want a PR.

**Reply by buchdag** [reply]
@SchoNie I'll try to read up on the subject by the end of the month, but it seems fairly different from DCT from the end user's perspective.
